File: ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProjectImpl.java
/*
GanttProject is an opensource project management tool.
Copyright (C) 2005-2011 Dmitry Barashev, GanttProject team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject;

import biz.ganttproject.core.calendar.GPCalendarCalc;
import biz.ganttproject.core.calendar.GPCalendarListener;
import biz.ganttproject.core.calendar.WeekendCalendarImpl;
import biz.ganttproject.core.option.ColorOption;
import biz.ganttproject.core.option.DefaultColorOption;
import biz.ganttproject.core.time.TimeUnitStack;
import biz.ganttproject.core.time.impl.GPTimeUnitStack;
import com.google.common.base.Strings;
import net.sourceforge.ganttproject.document.Document;
import net.sourceforge.ganttproject.document.DocumentManager;
import net.sourceforge.ganttproject.gui.NotificationManager;
import net.sourceforge.ganttproject.gui.UIConfiguration;
import net.sourceforge.ganttproject.gui.options.model.GP1XOptionConverter;
import net.sourceforge.ganttproject.language.GanttLanguage;
import net.sourceforge.ganttproject.resource.HumanResourceManager;
import net.sourceforge.ganttproject.roles.RoleManager;
import net.sourceforge.ganttproject.task.CustomColumnsManager;
import net.sourceforge.ganttproject.task.Task;
import net.sourceforge.ganttproject.task.TaskContainmentHierarchyFacade;
import net.sourceforge.ganttproject.task.TaskManager;
import net.sourceforge.ganttproject.task.TaskManagerConfig;

import java.awt.*;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class GanttProjectImpl implements IGanttProject {
  private static final GanttLanguage language = GanttLanguage.getInstance();
  private String myProjectName;
  private String myDescription;
  private String myOrganization;
  private String myWebLink;
  private final TaskManager myTaskManager;
  private final HumanResourceManager myResourceManager;
  private final TaskManagerConfigImpl myTaskManagerConfig;
  private Document myDocument;
  private final List<ProjectEventListener> myListeners = new ArrayList<ProjectEventListener>();
  private UIConfiguration myUIConfiguration;
  private final CustomColumnsManager myTaskCustomColumnManager;
  private final List<GanttPreviousState> myBaselines = new ArrayList<GanttPreviousState>();
  private final WeekendCalendarImpl myCalendar = new WeekendCalendarImpl();

  public GanttProjectImpl() {
    myResourceManager = new HumanResourceManager(RoleManager.Access.getInstance().getDefaultRole(),
        new CustomColumnsManager());
    myTaskManagerConfig = new TaskManagerConfigImpl(myResourceManager, myCalendar, GanttLanguage.getInstance());
    myTaskManager = TaskManager.Access.newInstance(null, myTaskManagerConfig);
    myUIConfiguration = new UIConfiguration(Color.BLUE, true);
    myTaskCustomColumnManager = new CustomColumnsManager();
    myCalendar.addListener(new GPCalendarListener() {
      @Override
      public void onCalendarChange() {
        setModified();
      }
    });
  }

  @Override
  public String getProjectName() {
    return myProjectName;
  }

  @Override
  public void setProjectName(String projectName) {
    myProjectName = projectName;
  }

  @Override
  public String getDescription() {
    return Strings.nullToEmpty(myDescription);
  }

  @Override
  public void setDescription(String description) {
    myDescription = description;
  }

  @Override
  public String getOrganization() {
    return myOrganization;
  }

  @Override
  public void setOrganization(String organization) {
    myOrganization = organization;
  }

  @Override
  public String getWebLink() {
    return myWebLink;
  }

  @Override
  public void setWebLink(String webLink) {
    myWebLink = webLink;
  }

  public Task newTask() {
    Task result = getTaskManager().createTask();
    getTaskManager().getTaskHierarchy().move(result, getTaskManager().getRootTask());
    return result;
  }

  public GanttLanguage getLanguage() {
    return language;
  }

  @Override
  public UIConfiguration getUIConfiguration() {
    return myUIConfiguration;
  }

  @Override
  public HumanResourceManager getHumanResourceManager() {
    return myResourceManager;
  }

  @Override
  public RoleManager getRoleManager() {
    return RoleManager.Access.getInstance();
  }

  @Override
  public TaskManager getTaskManager() {
    return myTaskManager;
  }

  @Override
  public TaskContainmentHierarchyFacade getTaskContainment() {
    return getTaskManager().getTaskHierarchy();
  }

  @Override
  public GPCalendarCalc getActiveCalendar() {
    return myTaskManagerConfig.getCalendar();
  }

  @Override
  public TimeUnitStack getTimeUnitStack() {
    return myTaskManagerConfig.getTimeUnitStack();
  }

  @Override
  public void setModified() {
    // TODO Auto-generated method stub
  }

  @Override
  public void setModified(boolean modified) {
    // TODO Auto-generated method stub
  }

  @Override
  public void close() {
    // TODO Auto-generated method stub
  }

  @Override
  public Document getDocument() {
    return myDocument;
  }

  @Override
  public void setDocument(Document document) {
    myDocument = document;
  }

  @Override
  public void addProjectEventListener(ProjectEventListener listener) {
    myListeners.add(listener);
  }

  @Override
  public void removeProjectEventListener(ProjectEventListener listener) {
    myListeners.remove(listener);
  }

  @Override
  public boolean isModified() {
    // TODO Auto-generated method stub
    return false;
  }

  @Override
  public void open(Document document) throws IOException {
    // TODO Auto-generated method stub
  }

  @Override
  public DocumentManager getDocumentManager() {
    // TODO Auto-generated method stub
    return null;
  }

  @Override
  public CustomPropertyManager getResourceCustomPropertyManager() {
    return myResourceManager.getCustomPropertyManager();
  }

  ;

  private static Color DEFAULT_TASK_COLOR = new Color(140, 182, 206);

  private static class TaskManagerConfigImpl implements TaskManagerConfig {
    private final HumanResourceManager myResourceManager;
    private final GPTimeUnitStack myTimeUnitStack;
    private final GPCalendarCalc myCalendar;
    private final ColorOption myDefaultTaskColorOption;

    private TaskManagerConfigImpl(HumanResourceManager resourceManager, GPCalendarCalc calendar, GanttLanguage i18n) {
      myResourceManager = resourceManager;
      myTimeUnitStack = new GPTimeUnitStack();
      myCalendar = calendar;
      myDefaultTaskColorOption = new DefaultTaskColorOption(DEFAULT_TASK_COLOR);
    }

    @Override
    public Color getDefaultColor() {
      return myDefaultTaskColorOption.getValue();
    }

    @Override
    public ColorOption getDefaultColorOption() {
      return myDefaultTaskColorOption;
    }

    @Override
    public GPCalendarCalc getCalendar() {
      return myCalendar;
    }

    @Override
    public TimeUnitStack getTimeUnitStack() {
      return myTimeUnitStack;
    }

    @Override
    public HumanResourceManager getResourceManager() {
      return myResourceManager;
    }

    @Override
    public URL getProjectDocumentURL() {
      return null;
    }

    @Override
    public NotificationManager getNotificationManager() {
      return null;
    }
  }

  @Override
  public CustomPropertyManager getTaskCustomColumnManager() {
    return myTaskCustomColumnManager;
  }

  @Override
  public List<GanttPreviousState> getBaselines() {
    return myBaselines;
  }

  public void repaintResourcePanel() {
    // TODO Auto-generated method stub
  }


  static class DefaultTaskColorOption extends DefaultColorOption implements GP1XOptionConverter {
    DefaultTaskColorOption() {
      this(DEFAULT_TASK_COLOR);
    }

    private DefaultTaskColorOption(Color defaultColor) {
      super("taskDefaultColor", defaultColor);
    }

    @Override
    public String getTagName() {
      return "colors";
    }

    @Override
    public String getAttributeName() {
      return "tasks";
    }

    @Override
    public void loadValue(String legacyValue) {
      loadPersistentValue(legacyValue);
      commit();
    }
  }


}

File: ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProjectImpl.kt
/*
GanttProject is an opensource project management tool.
Copyright (C) 2005-2021 Dmitry Barashev, GanttProject team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package net.sourceforge.ganttproject

import biz.ganttproject.app.SimpleBarrier
import biz.ganttproject.app.TimerBarrier
import biz.ganttproject.app.TwoPhaseBarrierImpl
import biz.ganttproject.core.calendar.GPCalendarCalc
import biz.ganttproject.core.calendar.ImportCalendarOption
import biz.ganttproject.core.calendar.WeekendCalendarImpl
import biz.ganttproject.core.option.BooleanOption
import biz.ganttproject.core.option.ColorOption
import biz.ganttproject.core.option.DefaultBooleanOption
import biz.ganttproject.core.option.DefaultColorOption
import biz.ganttproject.core.time.TimeUnitStack
import biz.ganttproject.core.time.impl.GPTimeUnitStack
import biz.ganttproject.customproperty.CustomPropertyManager
import biz.ganttproject.ganttview.TaskFilterManager
import net.sourceforge.ganttproject.document.Document
import net.sourceforge.ganttproject.document.DocumentManager
import net.sourceforge.ganttproject.gui.NotificationManager
import net.sourceforge.ganttproject.gui.UIConfiguration
import net.sourceforge.ganttproject.gui.UIFacade
import net.sourceforge.ganttproject.gui.options.model.GP1XOptionConverter
import net.sourceforge.ganttproject.importer.BufferProject
import net.sourceforge.ganttproject.importer.TaskMapping
import net.sourceforge.ganttproject.language.GanttLanguage
import net.sourceforge.ganttproject.resource.HumanResourceManager
import net.sourceforge.ganttproject.resource.HumanResourceMerger
import net.sourceforge.ganttproject.resource.OverwritingMerger
import net.sourceforge.ganttproject.roles.RoleManager
import net.sourceforge.ganttproject.storage.LazyProjectDatabaseProxy
import net.sourceforge.ganttproject.storage.ProjectDatabase
import net.sourceforge.ganttproject.task.*
import net.sourceforge.ganttproject.task.event.createTaskListenerWithTimerBarrier
import java.awt.Color
import java.io.IOException
import java.net.URL
import javax.swing.SwingUtilities

fun interface ErrorUi {
  fun show(ex: Exception)
}

open class GanttProjectImpl(taskManager: TaskManagerImpl? = null,
                            override val projectDatabase: ProjectDatabase = LazyProjectDatabaseProxy(
                              {error("Not supposed to be called")},
                              {error("Not supposed to be called")},
                            ))
  : IGanttProject {
  val listeners: MutableList<ProjectEventListener> = mutableListOf()
  override val baselines: MutableList<GanttPreviousState> = ArrayList()

  override var projectName: String = ""
  override var description: String = ""
  override var organization: String = ""
  override var webLink: String = ""

  val language: GanttLanguage get() = GanttLanguage.getInstance()
  private val myCalendar = WeekendCalendarImpl()
  final override val humanResourceManager = HumanResourceManager(
    RoleManager.Access.getInstance().defaultRole,
    CustomColumnsManager()
  )
  override val resourceCustomPropertyManager: CustomPropertyManager get() = humanResourceManager.customPropertyManager
  private val myTaskManagerConfig = TaskManagerConfigImpl(humanResourceManager, myCalendar)
  final override val taskManager: TaskManagerImpl = taskManager ?: TaskManagerImpl(null, myTaskManagerConfig)
  override val uIConfiguration = UIConfiguration(Color.BLUE, true)
  override val taskCustomColumnManager: CustomPropertyManager get() = taskManager.customPropertyManager
  override val taskFilterManager = TaskFilterManager(this.taskManager)
  override val roleManager: RoleManager
    get() = RoleManager.Access.getInstance()

  override var isModified: Boolean = false
  override val activeCalendar: GPCalendarCalc get() = myTaskManagerConfig.calendar
  override val timeUnitStack: TimeUnitStack get() = myTaskManagerConfig.timeUnitStack
  override var document: Document get() { TODO() } set(_) {
    TODO()
  }
  override val documentManager: DocumentManager
    get() = TODO("Not yet implemented")

  init {
    myCalendar.addListener { setModified() }
  }

  override fun setModified() {
    isModified = true
  }

  override fun close() {
    // TODO Auto-generated method stub
  }

  override fun addProjectEventListener(listener: ProjectEventListener) {
    listeners.add(listener)
  }

  override fun removeProjectEventListener(listener: ProjectEventListener) {
    listeners.remove(listener)
  }

  fun fireProjectModified(isModified: Boolean, errorUi: ErrorUi) {
    for (modifiedStateChangeListener in listeners) {
      try {
        if (isModified) {
          modifiedStateChangeListener.projectModified()
        } else {
          modifiedStateChangeListener.projectSaved()
        }
      } catch (e: Exception) {
        errorUi.show(e)
      }
    }
  }

  protected open fun fireProjectCreated() {
    for (modifiedStateChangeListener in listeners) {
      modifiedStateChangeListener.projectCreated()
    }
    // A new project just got created, so it is not yet modified
    SwingUtilities.invokeLater { isModified = false }
  }

  protected open fun fireProjectClosed() {
    for (modifiedStateChangeListener in listeners) {
      modifiedStateChangeListener.projectClosed()
    }
  }

  protected open fun fireProjectOpened() {
    val barrier = TwoPhaseBarrierImpl<IGanttProject>(this)
    for (l in listeners) {
      l.projectOpened(barrier, barrier)
    }
  }

  @Throws(Document.DocumentException::class, IOException::class)
  override fun restore(fromDocument: Document) {
    restoreProject(fromDocument, this.listeners)
  }

  @Throws(IOException::class)
  override fun open(document: Document) {
    // TODO Auto-generated method stub
  }

  override fun importProject(
    bufferProject: BufferProject,
    mergeOption: HumanResourceMerger.MergeResourcesOption,
    importCalendarOption: ImportCalendarOption?,
    closeCurrentProject: Boolean
  ): TaskMapping {
      roleManager.importData(bufferProject.roleManager)
      if (importCalendarOption != null) {
        activeCalendar.importCalendar(bufferProject.activeCalendar, importCalendarOption)
      }
      val that2thisResourceCustomDefs =
        resourceCustomPropertyManager.importData(bufferProject.resourceCustomPropertyManager)
      val original2ImportedResource = humanResourceManager.importData(
        bufferProject.humanResourceManager, OverwritingMerger(mergeOption), that2thisResourceCustomDefs
      )
      val that2thisCustomDefs = taskCustomColumnManager.importData(bufferProject.taskCustomColumnManager)
      val origTaskManager = taskManager
      try {
        origTaskManager.setEventsEnabled(false)
        val result = origTaskManager.importData(bufferProject.taskManager, that2thisCustomDefs)
        origTaskManager.importAssignments(
          bufferProject.taskManager, humanResourceManager,
          result, original2ImportedResource
        )
        return result
      } finally {
        origTaskManager.setEventsEnabled(true)
      }
  }
}

private val DEFAULT_TASK_COLOR = Color(140, 182, 206)

private class TaskManagerConfigImpl(
  private val myResourceManager: HumanResourceManager,
  calendar: GPCalendarCalc
) : TaskManagerConfig {
  private val myTimeUnitStack: GPTimeUnitStack
  private val myCalendar: GPCalendarCalc
  private val myDefaultTaskColorOption: ColorOption
  private val mySchedulerDisabledOption: BooleanOption
  override fun getDefaultColor(): Color {
    return myDefaultTaskColorOption.value!!
  }

  override fun getDefaultColorOption(): ColorOption {
    return myDefaultTaskColorOption
  }

  override fun getSchedulerDisabledOption(): BooleanOption {
    return mySchedulerDisabledOption
  }

  override fun getCalendar(): GPCalendarCalc {
    return myCalendar
  }

  override fun getTimeUnitStack(): TimeUnitStack {
    return myTimeUnitStack
  }

  override fun getResourceManager(): HumanResourceManager {
    return myResourceManager
  }

  override fun getProjectDocumentURL(): URL {
    TODO()
  }

  override fun getNotificationManager(): NotificationManager {
    TODO()
  }

  init {
    myTimeUnitStack = GPTimeUnitStack()
    myCalendar = calendar
    myDefaultTaskColorOption = DefaultTaskColorOption(DEFAULT_TASK_COLOR)
    mySchedulerDisabledOption = DefaultBooleanOption("scheduler.disabled", false)
  }
}

internal class DefaultTaskColorOption internal constructor(defaultColor: Color) :
  DefaultColorOption("taskDefaultColor", defaultColor), GP1XOptionConverter {
  constructor() : this(DEFAULT_TASK_COLOR)

  override fun getTagName(): String {
    return "colors"
  }

  override fun getAttributeName(): String {
    return "tasks"
  }

  override fun loadValue(legacyValue: String) {
    loadPersistentValue(legacyValue)
    commit()
  }
}

internal fun (IGanttProject).restoreProject(fromDocument: Document, listeners: List<ProjectEventListener>) {
  restoreProject(listeners) {
    fromDocument.read()
  }
}

internal fun <T> (IGanttProject).restoreProject(listeners: List<ProjectEventListener>, closeCurrentProject: Boolean = true, restoreCode: ()->T): T {
  val completionPromise = SimpleBarrier<Document>()
  listeners.forEach { it.projectRestoring(completionPromise) }
  val projectDocument = document
  if (closeCurrentProject) {
    close()
  }
  val algs = taskManager.algorithmCollection
  return try {
    algs.scheduler.isEnabled = false
    algs.recalculateTaskScheduleAlgorithm.isEnabled = false
    algs.adjustTaskBoundsAlgorithm.isEnabled = false
    restoreCode()
  } finally {
    algs.recalculateTaskScheduleAlgorithm.isEnabled = true
    algs.adjustTaskBoundsAlgorithm.isEnabled = true
    algs.scheduler.isEnabled = true
    completionPromise.resolve(projectDocument)
    document = projectDocument
  }
}


internal fun createProjectModificationListener(project: IGanttProject, uiFacade: UIFacade) =
  createTaskListenerWithTimerBarrier(timerBarrier = TimerBarrier(1000).apply {
    await { project.setModified() }
  }).also {
      it.taskAddedHandler = {
        project.setModified()
        uiFacade.viewIndex = UIFacade.GANTT_INDEX
        uiFacade.refresh()
      }
    }
